File: contracts/assembly/stablecoin.ts
import { storage, Context, contract, view, mutate, assert } from "./contract";

// Contract storage keys
const OWNER: string = "owner";
const TOTAL_SUPPLY: string = "totalSupply";
const BALANCES_PREFIX: string = "balance_";
const PAUSED: string = "paused";
const BLACKLIST_PREFIX: string = "blacklist_";
const MINTER_PREFIX: string = "minter_";
const COLLATERAL_PREFIX: string = "collateral_";
const COLLATERAL_TYPE_PREFIX: string = "collateral_type_";
const BINOM_TOKEN_ADDRESS: string = "binom_token_address";
const COLLATERAL_RATIO: string = "collateral_ratio";
const FIAT_RESERVE: string = "fiat_reserve";

// Collateral types
enum CollateralType {
    FIAT,
    BINOM
}

// Events
export function emitTransfer(from: string, to: string, amount: u64): void {
    // Event emission will be handled by the VM
}

export function emitMint(to: string, amount: u64, collateralType: CollateralType): void {
    // Event emission will be handled by the VM
}

export function emitBurn(from: string, amount: u64): void {
    // Event emission will be handled by the VM
}

export function emitCollateralAdded(from: string, amount: u64, collateralType: CollateralType): void {
    // Event emission will be handled by the VM
}

export function emitCollateralRemoved(to: string, amount: u64, collateralType: CollateralType): void {
    // Event emission will be handled by the VM
}

@contract
export class PaperDollar {
    // Initialize the contract
    constructor() {
        if (!storage.get<string>(OWNER, "")) {
            storage.set<string>(OWNER, Context.caller);
            storage.set<u64>(TOTAL_SUPPLY, 0);
            storage.set<boolean>(PAUSED, false);
            storage.set<u64>(COLLATERAL_RATIO, 150); // 150% collateralization ratio
            storage.set<u64>(FIAT_RESERVE, 0);
        }
    }

    // Only owner modifier
    private checkOwner(): void {
        assert(storage.get<string>(OWNER, "") == Context.caller, "Only owner can call this function");
    }

    // Not paused modifier
    private checkNotPaused(): void {
        assert(!storage.get<boolean>(PAUSED, false), "Contract is paused");
    }

    // Not blacklisted modifier
    private checkNotBlacklisted(address: string): void {
        assert(!this.isBlacklisted(address), "Address is blacklisted");
    }

    // Get balance of an address
    @view
    getBalance(address: string): u64 {
        return storage.get<u64>(BALANCES_PREFIX + address, 0);
    }

    // Get total supply
    @view
    getTotalSupply(): u64 {
        return storage.get<u64>(TOTAL_SUPPLY, 0);
    }

    // Get collateral balance
    @view
    getCollateralBalance(address: string, collateralType: CollateralType): u64 {
        return storage.get<u64>(COLLATERAL_PREFIX + collateralType.toString() + "_" + address, 0);
    }

    // Get collateral type
    @view
    getCollateralType(address: string): CollateralType {
        return storage.get<CollateralType>(COLLATERAL_TYPE_PREFIX + address, CollateralType.FIAT);
    }

    // Get collateral ratio
    @view
    getCollateralRatio(): u64 {
        return storage.get<u64>(COLLATERAL_RATIO, 150);
    }

    // Get fiat reserve
    @view
    getFiatReserve(): u64 {
        return storage.get<u64>(FIAT_RESERVE, 0);
    }

    // Check if address is blacklisted
    @view
    isBlacklisted(address: string): boolean {
        return storage.get<boolean>(BLACKLIST_PREFIX + address, false);
    }

    // Check if address is a minter
    @view
    isMinter(address: string): boolean {
        return storage.get<boolean>(MINTER_PREFIX + address, false);
    }

    // Set BINOM token address (only owner)
    @mutate
    setBinomTokenAddress(address: string): boolean {
        this.checkOwner();
        storage.set<string>(BINOM_TOKEN_ADDRESS, address);
        return true;
    }

    // Set collateral ratio (only owner)
    @mutate
    setCollateralRatio(ratio: u64): boolean {
        this.checkOwner();
        assert(ratio >= 100, "Collateral ratio must be at least 100%");
        storage.set<u64>(COLLATERAL_RATIO, ratio);
        return true;
    }

    // Add collateral
    @mutate
    addCollateral(amount: u64, collateralType: CollateralType): boolean {
        const from = Context.caller;
        this.checkNotPaused();
        this.checkNotBlacklisted(from);

        if (collateralType == CollateralType.FIAT) {
            storage.set<u64>(FIAT_RESERVE, this.getFiatReserve() + amount);
        }

        const currentCollateral = this.getCollateralBalance(from, collateralType);
        storage.set<u64>(COLLATERAL_PREFIX + collateralType.toString() + "_" + from, currentCollateral + amount);
        storage.set<CollateralType>(COLLATERAL_TYPE_PREFIX + from, collateralType);

        emitCollateralAdded(from, amount, collateralType);
        return true;
    }

    // Remove collateral
    @mutate
    removeCollateral(amount: u64): boolean {
        const from = Context.caller;
        this.checkNotPaused();
        this.checkNotBlacklisted(from);

        const collateralType = this.getCollateralType(from);
        const currentCollateral = this.getCollateralBalance(from, collateralType);
        assert(currentCollateral >= amount, "Insufficient collateral");

        // Check if removing collateral would break the collateral ratio
        const userBalance = this.getBalance(from);
        const remainingCollateral = currentCollateral - amount;
        assert(remainingCollateral * 100 >= userBalance * this.getCollateralRatio(), "Collateral ratio would be too low");

        if (collateralType == CollateralType.FIAT) {
            storage.set<u64>(FIAT_RESERVE, this.getFiatReserve() - amount);
        }

        storage.set<u64>(COLLATERAL_PREFIX + collateralType.toString() + "_" + from, remainingCollateral);

        emitCollateralRemoved(from, amount, collateralType);
        return true;
    }

    // Transfer tokens
    @mutate
    transfer(to: string, amount: u64): boolean {
        const from = Context.caller;
        this.checkNotPaused();
        this.checkNotBlacklisted(from);
        this.checkNotBlacklisted(to);

        const fromBalance = this.getBalance(from);
        assert(fromBalance >= amount, "Insufficient balance");

        storage.set<u64>(BALANCES_PREFIX + from, fromBalance - amount);
        storage.set<u64>(BALANCES_PREFIX + to, this.getBalance(to) + amount);

        emitTransfer(from, to, amount);
        return true;
    }

    // Mint new tokens (only minters)
    @mutate
    mint(to: string, amount: u64): boolean {
        assert(this.isMinter(Context.caller), "Only minters can mint");
        this.checkNotPaused();
        this.checkNotBlacklisted(to);

        const collateralType = this.getCollateralType(to);
        const collateral = this.getCollateralBalance(to, collateralType);
        const currentBalance = this.getBalance(to);
        const newBalance = currentBalance + amount;

        // Check collateral ratio
        assert(collateral * 100 >= newBalance * this.getCollateralRatio(), "Insufficient collateral");

        const newTotalSupply = this.getTotalSupply() + amount;
        storage.set<u64>(TOTAL_SUPPLY, newTotalSupply);
        storage.set<u64>(BALANCES_PREFIX + to, newBalance);

        emitMint(to, amount, collateralType);
        return true;
    }

    // Burn tokens
    @mutate
    burn(amount: u64): boolean {
        const from = Context.caller;
        this.checkNotPaused();
        this.checkNotBlacklisted(from);

        const fromBalance = this.getBalance(from);
        assert(fromBalance >= amount, "Insufficient balance");

        storage.set<u64>(BALANCES_PREFIX + from, fromBalance - amount);
        storage.set<u64>(TOTAL_SUPPLY, this.getTotalSupply() - amount);

        emitBurn(from, amount);
        return true;
    }

    // Add a minter (only owner)
    @mutate
    addMinter(address: string): boolean {
        this.checkOwner();
        storage.set<boolean>(MINTER_PREFIX + address, true);
        return true;
    }

    // Remove a minter (only owner)
    @mutate
    removeMinter(address: string): boolean {
        this.checkOwner();
        storage.set<boolean>(MINTER_PREFIX + address, false);
        return true;
    }

    // Blacklist an address (only owner)
    @mutate
    blacklist(address: string): boolean {
        this.checkOwner();
        storage.set<boolean>(BLACKLIST_PREFIX + address, true);
        return true;
    }

    // Remove address from blacklist (only owner)
    @mutate
    unblacklist(address: string): boolean {
        this.checkOwner();
        storage.set<boolean>(BLACKLIST_PREFIX + address, false);
        return true;
    }

    // Pause the contract (only owner)
    @mutate
    pause(): boolean {
        this.checkOwner();
        storage.set<boolean>(PAUSED, true);
        return true;
    }

    // Unpause the contract (only owner)
    @mutate
    unpause(): boolean {
        this.checkOwner();
        storage.set<boolean>(PAUSED, false);
        return true;
    }

    // Transfer ownership (only owner)
    @mutate
    transferOwnership(newOwner: string): boolean {
        this.checkOwner();
        storage.set<string>(OWNER, newOwner);
        return true;
    }
} 

File: contracts/stablecoin/src/stablecoin.rs
use crate::storage::{Storage, StorageKeys};
use crate::context::Context;
use crate::events::*;
use crate::errors::*;
use crate::{require, safe_add, safe_sub, safe_mul};

pub struct PaperDollar;

impl PaperDollar {
    /// Initialize the contract
    pub fn new() -> Self {
        // Set the caller as the owner if no owner exists
        if !Storage::exists(StorageKeys::OWNER) {
            let caller = Context::caller();
            Storage::set_string(StorageKeys::OWNER, &caller);
            Storage::set_u64(StorageKeys::TOTAL_SUPPLY, 0);
            Storage::set_bool(StorageKeys::PAUSED, false);
            Storage::set_u64(StorageKeys::COLLATERAL_RATIO, 150); // Default 150%
            Storage::set_u64(StorageKeys::FIAT_RESERVE, 0);
        }
        
        PaperDollar
    }

    // Modifier functions for access control
    fn check_owner(&self) -> ContractResult<()> {
        let owner = Storage::get_string(StorageKeys::OWNER, "");
        require!(owner == Context::caller(), ContractError::OnlyOwner);
        Ok(())
    }

    fn check_not_paused(&self) -> ContractResult<()> {
        require!(!Storage::get_bool(StorageKeys::PAUSED, false), ContractError::ContractPaused);
        Ok(())
    }

    fn check_not_blacklisted(&self, address: &str) -> ContractResult<()> {
        require!(!self.is_blacklisted(address), ContractError::AddressBlacklisted);
        Ok(())
    }

    fn check_minter(&self) -> ContractResult<()> {
        require!(self.is_minter(&Context::caller()), ContractError::OnlyMinters);
        Ok(())
    }

    // View functions
    pub fn get_balance(&self, address: &str) -> u64 {
        let key = format!("{}{}", StorageKeys::BALANCES_PREFIX, address);
        Storage::get_u64(&key, 0)
    }

    pub fn get_total_supply(&self) -> u64 {
        Storage::get_u64(StorageKeys::TOTAL_SUPPLY, 0)
    }

    pub fn get_collateral_balance(&self, address: &str, collateral_type: CollateralType) -> u64 {
        let key = format!("{}{}_{}",
            StorageKeys::COLLATERAL_PREFIX,
            collateral_type.to_string(),
            address
        );
        Storage::get_u64(&key, 0)
    }

    pub fn get_collateral_type(&self, address: &str) -> CollateralType {
        let key = format!("{}{}", StorageKeys::COLLATERAL_TYPE_PREFIX, address);
        let value = Storage::get_u64(&key, 0);
        CollateralType::from(value as u8)
    }

    pub fn get_collateral_ratio(&self) -> u64 {
        Storage::get_u64(StorageKeys::COLLATERAL_RATIO, 150)
    }

    pub fn get_fiat_reserve(&self) -> u64 {
        Storage::get_u64(StorageKeys::FIAT_RESERVE, 0)
    }

    pub fn is_blacklisted(&self, address: &str) -> bool {
        let key = format!("{}{}", StorageKeys::BLACKLIST_PREFIX, address);
        Storage::get_bool(&key, false)
    }

    pub fn is_minter(&self, address: &str) -> bool {
        let key = format!("{}{}", StorageKeys::MINTER_PREFIX, address);
        Storage::get_bool(&key, false)
    }

    pub fn get_owner(&self) -> String {
        Storage::get_string(StorageKeys::OWNER, "")
    }

    pub fn is_paused(&self) -> bool {
        Storage::get_bool(StorageKeys::PAUSED, false)
    }

    // Mutating functions

    /// Set BINOM token address (only owner)
    pub fn set_binom_token_address(&mut self, address: &str) -> ContractResult<bool> {
        self.check_owner()?;
        Storage::set_string(StorageKeys::BINOM_TOKEN_ADDRESS, address);
        Ok(true)
    }

    /// Set collateral ratio (only owner)
    pub fn set_collateral_ratio(&mut self, ratio: u64) -> ContractResult<bool> {
        self.check_owner()?;
        require!(ratio >= 100, ContractError::InvalidCollateralRatio);
        Storage::set_u64(StorageKeys::COLLATERAL_RATIO, ratio);
        Ok(true)
    }

    /// Add collateral
    pub fn add_collateral(&mut self, amount: u64, collateral_type: CollateralType) -> ContractResult<bool> {
        let from = Context::caller();
        require!(amount > 0, ContractError::ZeroAmount);
        
        self.check_not_paused()?;
        self.check_not_blacklisted(&from)?;

        if collateral_type == CollateralType::Fiat {
            let new_reserve = safe_add(self.get_fiat_reserve(), amount)?;
            Storage::set_u64(StorageKeys::FIAT_RESERVE, new_reserve);
        }

        let current_collateral = self.get_collateral_balance(&from, collateral_type);
        let new_collateral = safe_add(current_collateral, amount)?;
        
        let collateral_key = format!("{}{}_{}",
            StorageKeys::COLLATERAL_PREFIX,
            collateral_type.to_string(),
            from
        );
        Storage::set_u64(&collateral_key, new_collateral);

        let type_key = format!("{}{}", StorageKeys::COLLATERAL_TYPE_PREFIX, from);
        Storage::set_u64(&type_key, collateral_type as u64);

        emit_collateral_added(&from, amount, collateral_type);
        Ok(true)
    }

    /// Remove collateral
    pub fn remove_collateral(&mut self, amount: u64) -> ContractResult<bool> {
        let from = Context::caller();
        require!(amount > 0, ContractError::ZeroAmount);
        
        self.check_not_paused()?;
        self.check_not_blacklisted(&from)?;

        let collateral_type = self.get_collateral_type(&from);
        let current_collateral = self.get_collateral_balance(&from, collateral_type);
        require!(current_collateral >= amount, ContractError::InsufficientCollateral);

        // Check if removing collateral would break the collateral ratio
        let user_balance = self.get_balance(&from);
        let remaining_collateral = safe_sub(current_collateral, amount)?;
        let required_collateral = safe_mul(user_balance, self.get_collateral_ratio())?;
        let collateral_value = safe_mul(remaining_collateral, 100)?;
        
        require!(collateral_value >= required_collateral, ContractError::CollateralRatioTooLow);

        if collateral_type == CollateralType::Fiat {
            let new_reserve = safe_sub(self.get_fiat_reserve(), amount)?;
            Storage::set_u64(StorageKeys::FIAT_RESERVE, new_reserve);
        }

        let collateral_key = format!("{}{}_{}",
            StorageKeys::COLLATERAL_PREFIX,
            collateral_type.to_string(),
            from
        );
        Storage::set_u64(&collateral_key, remaining_collateral);

        emit_collateral_removed(&from, amount, collateral_type);
        Ok(true)
    }

    /// Transfer tokens
    pub fn transfer(&mut self, to: &str, amount: u64) -> ContractResult<bool> {
        let from = Context::caller();
        require!(amount > 0, ContractError::ZeroAmount);
        require!(from != to, ContractError::TransferToSelf);
        
        self.check_not_paused()?;
        self.check_not_blacklisted(&from)?;
        self.check_not_blacklisted(to)?;

        let from_balance = self.get_balance(&from);
        require!(from_balance >= amount, ContractError::InsufficientBalance);

        let new_from_balance = safe_sub(from_balance, amount)?;
        let to_balance = self.get_balance(to);
        let new_to_balance = safe_add(to_balance, amount)?;

        let from_key = format!("{}{}", StorageKeys::BALANCES_PREFIX, from);
        let to_key = format!("{}{}", StorageKeys::BALANCES_PREFIX, to);
        
        Storage::set_u64(&from_key, new_from_balance);
        Storage::set_u64(&to_key, new_to_balance);

        emit_transfer(&from, to, amount);
        Ok(true)
    }

    /// Mint new tokens (only minters)
    pub fn mint(&mut self, to: &str, amount: u64) -> ContractResult<bool> {
        require!(amount > 0, ContractError::ZeroAmount);
        
        self.check_minter()?;
        self.check_not_paused()?;
        self.check_not_blacklisted(to)?;

        let collateral_type = self.get_collateral_type(to);
        let collateral = self.get_collateral_balance(to, collateral_type);
        let current_balance = self.get_balance(to);
        let new_balance = safe_add(current_balance, amount)?;

        // Check collateral ratio
        let required_collateral = safe_mul(new_balance, self.get_collateral_ratio())?;
        let collateral_value = safe_mul(collateral, 100)?;
        require!(collateral_value >= required_collateral, ContractError::InsufficientCollateral);

        let new_total_supply = safe_add(self.get_total_supply(), amount)?;
        Storage::set_u64(StorageKeys::TOTAL_SUPPLY, new_total_supply);

        let balance_key = format!("{}{}", StorageKeys::BALANCES_PREFIX, to);
        Storage::set_u64(&balance_key, new_balance);

        emit_mint(to, amount, collateral_type);
        Ok(true)
    }

    /// Burn tokens
    pub fn burn(&mut self, amount: u64) -> ContractResult<bool> {
        let from = Context::caller();
        require!(amount > 0, ContractError::ZeroAmount);
        
        self.check_not_paused()?;
        self.check_not_blacklisted(&from)?;

        let from_balance = self.get_balance(&from);
        require!(from_balance >= amount, ContractError::InsufficientBalance);

        let new_balance = safe_sub(from_balance, amount)?;
        let new_total_supply = safe_sub(self.get_total_supply(), amount)?;

        let balance_key = format!("{}{}", StorageKeys::BALANCES_PREFIX, from);
        Storage::set_u64(&balance_key, new_balance);
        Storage::set_u64(StorageKeys::TOTAL_SUPPLY, new_total_supply);

        emit_burn(&from, amount);
        Ok(true)
    }

    /// Add a minter (only owner)
    pub fn add_minter(&mut self, address: &str) -> ContractResult<bool> {
        self.check_owner()?;
        let key = format!("{}{}", StorageKeys::MINTER_PREFIX, address);
        Storage::set_bool(&key, true);
        emit_minter_added(address);
        Ok(true)
    }

    /// Remove a minter (only owner)
    pub fn remove_minter(&mut self, address: &str) -> ContractResult<bool> {
        self.check_owner()?;
        let key = format!("{}{}", StorageKeys::MINTER_PREFIX, address);
        Storage::set_bool(&key, false);
        emit_minter_removed(address);
        Ok(true)
    }

    /// Blacklist an address (only owner)
    pub fn blacklist(&mut self, address: &str) -> ContractResult<bool> {
        self.check_owner()?;
        let key = format!("{}{}", StorageKeys::BLACKLIST_PREFIX, address);
        Storage::set_bool(&key, true);
        emit_blacklist_added(address);
        Ok(true)
    }

    /// Remove address from blacklist (only owner)
    pub fn unblacklist(&mut self, address: &str) -> ContractResult<bool> {
        self.check_owner()?;
        let key = format!("{}{}", StorageKeys::BLACKLIST_PREFIX, address);
        Storage::set_bool(&key, false);
        emit_blacklist_removed(address);
        Ok(true)
    }

    /// Pause the contract (only owner)
    pub fn pause(&mut self) -> ContractResult<bool> {
        self.check_owner()?;
        Storage::set_bool(StorageKeys::PAUSED, true);
        emit_paused();
        Ok(true)
    }

    /// Unpause the contract (only owner)
    pub fn unpause(&mut self) -> ContractResult<bool> {
        self.check_owner()?;
        Storage::set_bool(StorageKeys::PAUSED, false);
        emit_unpaused();
        Ok(true)
    }

    /// Transfer ownership (only owner)
    pub fn transfer_ownership(&mut self, new_owner: &str) -> ContractResult<bool> {
        self.check_owner()?;
        let old_owner = self.get_owner();
        Storage::set_string(StorageKeys::OWNER, new_owner);
        emit_ownership_transferred(&old_owner, new_owner);
        Ok(true)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::with_caller;

    fn setup_contract() -> PaperDollar {
        Storage::clear();
        Context::reset();
        with_caller!("owner", {
            PaperDollar::new()
        })
    }

    #[test]
    fn test_initialization() {
        let contract = setup_contract();
        assert_eq!(contract.get_owner(), "owner");
        assert_eq!(contract.get_total_supply(), 0);
        assert_eq!(contract.get_collateral_ratio(), 150);
        assert!(!contract.is_paused());
    }

    #[test]
    fn test_add_collateral() {
        let mut contract = setup_contract();
        
        with_caller!("user1", {
            let result = contract.add_collateral(1000, CollateralType::Fiat);
            assert!(result.is_ok());
            assert_eq!(contract.get_collateral_balance("user1", CollateralType::Fiat), 1000);
            assert_eq!(contract.get_fiat_reserve(), 1000);
        });
    }

    #[test]
    fn test_mint_with_collateral() {
        let mut contract = setup_contract();
        
        // Add minter
        with_caller!("owner", {
            contract.add_minter("minter1").unwrap();
        });

        // User adds collateral
        with_caller!("user1", {
            contract.add_collateral(1500, CollateralType::Fiat).unwrap();
        });

        // Minter mints tokens
        with_caller!("minter1", {
            let result = contract.mint("user1", 1000);
            assert!(result.is_ok());
            assert_eq!(contract.get_balance("user1"), 1000);
            assert_eq!(contract.get_total_supply(), 1000);
        });
    }

    #[test]
    fn test_transfer() {
        let mut contract = setup_contract();
        
        // Setup: mint tokens to user1
        with_caller!("owner", {
            contract.add_minter("minter1").unwrap();
        });

        with_caller!("user1", {
            contract.add_collateral(1500, CollateralType::Fiat).unwrap();
        });

        with_caller!("minter1", {
            contract.mint("user1", 1000).unwrap();
        });

        // Transfer tokens
        with_caller!("user1", {
            let result = contract.transfer("user2", 500);
            assert!(result.is_ok());
            assert_eq!(contract.get_balance("user1"), 500);
            assert_eq!(contract.get_balance("user2"), 500);
        });
    }

    #[test]
    fn test_burn() {
        let mut contract = setup_contract();
        
        // Setup: mint tokens to user1
        with_caller!("owner", {
            contract.add_minter("minter1").unwrap();
        });

        with_caller!("user1", {
            contract.add_collateral(1500, CollateralType::Fiat).unwrap();
        });

        with_caller!("minter1", {
            contract.mint("user1", 1000).unwrap();
        });

        // Burn tokens
        with_caller!("user1", {
            let result = contract.burn(300);
            assert!(result.is_ok());
            assert_eq!(contract.get_balance("user1"), 700);
            assert_eq!(contract.get_total_supply(), 700);
        });
    }

    #[test]
    fn test_access_control() {
        let mut contract = setup_contract();
        
        // Non-owner cannot add minter
        with_caller!("user1", {
            let result = contract.add_minter("user2");
            assert!(matches!(result.unwrap_err(), ContractError::OnlyOwner));
        });

        // Non-minter cannot mint
        with_caller!("user1", {
            let result = contract.mint("user2", 100);
            assert!(matches!(result.unwrap_err(), ContractError::OnlyMinters));
        });
    }

    #[test]
    fn test_pause_functionality() {
        let mut contract = setup_contract();
        
        // Owner pauses contract
        with_caller!("owner", {
            contract.pause().unwrap();
            assert!(contract.is_paused());
        });

        // Operations should fail when paused
        with_caller!("user1", {
            let result = contract.add_collateral(1000, CollateralType::Fiat);
            assert!(matches!(result.unwrap_err(), ContractError::ContractPaused));
        });

        // Owner unpauses contract
        with_caller!("owner", {
            contract.unpause().unwrap();
            assert!(!contract.is_paused());
        });

        // Operations should work again
        with_caller!("user1", {
            let result = contract.add_collateral(1000, CollateralType::Fiat);
            assert!(result.is_ok());
        });
    }
} 